### ISSUE TITLE
Race condition in enqueue an operation

### ISSUE THREAD

**Issue by soareschen** [w3c/webrtc-pc#1218]
In step 6 of enqueue an operation:

> 6. If the length of [[operations]] is exactly 1, execute operation.

This result in race condition with synchronous code following call to RTCPeerConnection methods. That is, if length of the operation queue is 1, the operation is executed _before_ the next line of synchronous code; But if the length of operation queue is > 1, the operation is executed _after_ the current microtask return to the event loop. This can result in code having unexpected behavior depending on whether the internal operation queue is empty or not when the operation is enqueued.

For example, if the next line of code is `peerconnection.close()`, then the operation may or may not be executed depending on the operation queue length.

To fix this, the operation should be enqueued to the next tick if the length of operation queue is 1.

**Comment by taylor-b** [comment]
Thanks for bringing up this race condition; it came up in a previous discussion but we must have not filed an independent issue for it.

I agree it could be fixed by just saying "queue a task to execute *operation*".

**Comment by burnburn** [comment]
@jan-ivar can you take a look?

**Comment by jan-ivar** [comment]
**TL;DR;** The operations queue is legacy tech; it's more important that it operates as if it's not there in the typical case (empty queue).

@soareschen We discussed this ad nauseum [3 years ago](https://lists.w3.org/Archives/Public/public-webrtc/2015May/0064.html) to arrive at the logic we have now. I don't think anyone wishes to reopen this topic.

The operations queue is legacy tech. In other words, well-behaved code does not rely on it:
```js
await pc.setRemoteDescription(incoming.sdp);
await pc.createAnswer()); // not relying on queue
```
Only old-style broken code relies on it: 
```js
pc.setRemoteDescription(incoming.sdp);
pc.createAnswer(); // relies on queue
```
The old code should already be considered "broken" because it does not propagate errors correctly. It also has the races you observe.

That means we put more value in preserving the synchronous nature of the synchronous halves of the operations (as if the operations queue did not exist) rather than worry about races when the queue is used.

> For example, if the next line of code is peerconnection.close(), then the operation may or may not be executed depending on the operation queue length.

Not so, because all enqueued operations are defined to stall on `close`.

**Comment by taylor-b** [comment]
If this is broken code, and we don't care about it having race conditions, then why did we care about [the `createOffer(); addTrack();` race condition](https://github.com/w3c/webrtc-pc/issues/782)?

Also, if an application is doing `await pc.setRemoteDescription(incoming.sdp);`, why does it care if the "synchronous part" of the operation happens immediately or via a queued task?

**Comment by jan-ivar** [comment]
@taylor-b Because we can write non-broken code where it matters:
```js
let haveOffer = pc.createOffer();
pc.addTrack(track, stream);
await haveOffer;
```

> why does it care if the "synchronous part" of the operation happens immediately

WebIDL does all input validation synchronously, which means we can step over an async function in the JS debugger:
```js
let p = pc.setRemoteDescription(x);
await p;
```
...and immediately inspect the return value to see if the asynchronous operation got launched or not:
```js
p: Promise { "pending" }
```

`"pending"` tells us the asynchronous operation was successfully launched. It may fail later for some runtime reason, but it got off the ground (i.e. we at least crossed all the t's and dotted the i's correctly).
```js
p: Promise { "rejected" }
```
`"rejected"` suggests there was a problem with the inputs, most likely `TypeError` on `x`, and the asynchronous part didn't even get off the ground. I find this helpful when debugging JS.

Where WebIDL falls short, and we need prose to add extra input validation, it makes sense to follow this pattern, both for consistency, and because WebIDL may grow new abilities later, and we may switch to them in the future without side-effects.

We've adopted this approach in both mediacapture and WebRTC.

**Comment by soareschen** [comment]
@jan-ivar Thank you for your explanation. Now I understand much better on why the enqueue operation steps are described the way they are.

The way I see it is that the main difference the patch in  #1397 make is that the synchronous part of the operation also execute asynchronously if the operation queue length is 1. 

Currently in the spec, there are only two places where operations have synchronous part:

- `createAnswer` step 4.1
- `addIceCandidate` step 4

Both of the synchronous parts only concern with returning rejected promise synchronously in case that operation queue length is 1. As a result, the only observable side effect for the special case of length 1 is:

1. When `close()` is called in next synchronous step, promise may or may not resolve depending on operation queue length.
2. When immediately printing returned promise to console, the promise is shown as "rejected" or "pending" depending on operation queue length.

I find the second use case of console debugging to be weird. In Chrome inspector one can click an arrow cursor to expand objects including promises, and the expanded attributes show the latest state of the promise, which in this case should be "rejected" even if the initial state is "pending".  This seems more like a compensation for other browsers like Firefox that do not have the interactive feature to quickly inspect latest attributes of objects. 

But Chrome and Firefox also show rejected promises returned from failed cases like `createAnswer()` as pending promise, so I am more confused of which browser that debugging technique is being used. Note that the use case is also limited to implementation-specific debugging because it is impossible to query the current state of promises in JavaScript, which also make this untestable.

Let me try to understand your points with some pseudo-tests here. Based on current behavior, enqueue would execute operation synchronously if operation queue length is 1. So:

```javascript
// This should fulfill with answer
pc.setRemoteDescription(offer);
pc.createAnswer().then(() => console.log('pass'))

// This should reject
// Note that this test fail in current browsers
const promise = pc.createAnswer();
pc.close();
promise.catch(() => console.log('pass'));

// This should never resolve
pc.setRemoteDescription(offer); // increase queue length
const promise = pc.createAnswer();
pc.close();
promise.catch(() => console.log('fail'));

// This should also never resolve
pc.addIceCandidate(...); // rejects
const promise = pc.createAnswer();
pc.close();
promise.catch(() => console.log('fail'));

// This should show rejected promise in console
console.log('rejected promise:', pc.createAnswer());

// This should show pending promise in console
// that later have an updated rejected state
pc.createAnswer(); // rejects
console.log('pending promise:', pc.createAnswer());

// This should not have added track in offer
// because createOffer only have async part of operation
const promise = pc.createOffer();
pc.addTransceiver('audio');
promise.then(offer => console.log('offer should not have track:', offer);

// This should also not have added track in offer
pc.setLocalDescription(oldOffer); // increase queue length
const promise = pc.createOffer();
pc.addTransceiver('audio');
promise.then(offer => console.log('offer should not have track:', offer);
```

I think we can solve this with a test driven approach. I am away next week. After I'm back I plan to write test cases that match the intended behavior in the spec.

**Comment by jan-ivar** [comment]
@soareschen Thanks for finding the cases where this matters: not validation errors, but the state errors you mention (The spec has improved in the years since last time this came up, moving validation outside of the enqueuing. We can't do the same for state checks unfortunately).

> the only observable side effect for the special case of length 1 is

As designed, length 1 is the *typical* case, not a special case. This is central to my point.

> 1. When close() is called in next synchronous step, promise may or may not resolve depending on operation queue length.

Again, it neither resolves nor rejects after `close()`. Look at [enqueue](http://w3c.github.io/webrtc-pc/#enqueue-an-operation):

 * *"7. Upon fulfillment or rejection of the promise returned by the operation, run the following:*
   * *1. If connection's [[isClosed]] slot is true, abort these steps."*

> 2. When immediately printing returned promise to console, the promise is shown as "rejected" or "pending" depending on operation queue length.
> 
> I find the second use case of console debugging to be weird.

The queue is inherently weird, so this is correct behavior, a clue to the weirdness that just happened. Avoid the queue at peril, and there's nothing weird here. I'm averse to polluting the typical case with the weirdness of the atypical legacy case merely for consistency.

Your interactions with Firefox don't match mine. Breakpoint after `createAnswer()` in [this fiddle](https://fiddle.jshell.net/jib1/fsoL1rab/show/):
```js
▼ p: Promise { "rejected" }
    <state>: "rejected"
  ▶ <reason>: InvalidStateError
  ▶ __proto__: PromiseProto
```

> I think we can solve this with a test driven approach.

If you're suggesting the spec should change based on test results, then I find that backwards. Tests exist to file bugs against browsers, not the other way around. Implementations have too many bugs atm. In any case, I don't think a legacy accommodation should wag the dog.

In any case, this test is wrong because it should never resolve nor reject according to the spec:
```js
// This should reject
// Note that this test fail in current browsers
const promise = pc.createAnswer();
pc.close();
promise.catch(() => console.log('pass'));
```

These two tests are wrong because https://github.com/w3c/webrtc-pc/pull/875 says they should be included:
```js
// This should not have added track in offer
// because createOffer only have async part of operation
const promise = pc.createOffer();
pc.addTransceiver('audio');
promise.then(offer => console.log('offer should not have track:', offer);

// This should also not have added track in offer
pc.setLocalDescription(oldOffer); // increase queue length
const promise = pc.createOffer();
pc.addTransceiver('audio');
promise.then(offer => console.log('offer should not have track:', offer);
```

**Comment by jan-ivar** [comment]
In fact the very last test shows why no-one should rely on the queue. `addTransceiver` here is effectively called *before* the enqueued part of `createOffer`, even before the enqueued part of `setLocalDescription`! We found no easy way to keep it out of createOffer, short of tracking parallel state around, nor is it intuitive that it should be kept out. So we didn't, instead changing `createOffer`/`createAnswer` to pick up on changes.

The queue is a mess, and I'm not sure why we're digging this up again.

**Comment by taylor-b** [comment]
Let's actually examine what the synchronous steps we're talking about are:

1. Checking if `remoteDescription` is `null` in `createAnswer` and `addIceCandidate`.
2. Checking if an ICE candidate's `sdpMid`/`sdpMLineIndex` are non-null.
3. Checking that an ICE candidate's `sdpMid` matches an "m=" section.

I don't think these are errors you'd need a debugger to help figure out; they're pretty obvious.

But on the other hand, I also don't see any race conditions these steps could introduce; they're dependent only on the results of other enqueued operations. So the only possible race is with `pc.close()`, as pointed out at the top of this issue. Which is probably not worth fixing.

So, although I don't see the point of the "execute immediately if queue is empty" rule, it doesn't seem to really be hurting anything.

**Comment by jan-ivar** [comment]
And since the queue basically aborts on `pc.close()` that race seems moot.

**Comment by taylor-b** [comment]
Can we close this issue then (and the corresponding PR)?

**Comment by soareschen** [comment]
Thanks @jan-ivar for the explanation. This issue was initially created to seek clarification, and I am glad to learn that the steps are crafted for very specific reasons mentioned.

Closing this issue now and I will write tests that match the legacy behavior examples that we discussed.
